lib/synapse/service_watcher/base.rb

```diff
@@ -95,13 +100,44 @@ def validate_discovery_opts
     end
 
     def set_backends(new_backends)
-      if @keep_default_servers
-        @backends = @default_servers + new_backends
+      # Aggregate and deduplicate all potential backend service instances.
```

Comment:
I'm concerned with duplicates in the set of backends we write out to HAProxy in general. Like, we had a few cases where duplicate backends got written out and then synapse would only down a single one when we failed healthchecks.

If we dedup the default_servers separately doesn't that mean that we might end up with duplicates between the defaults and new_backends and could end up with the bad situation I mention above?